### PR TITLE
ddl: introduce ErrAutoIncrementOverflow to handle auto_increment overflow.

### LINKED ISSUE #34142: Unexpected side-effects of setting auto_increment
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
create table test1(
 a bigint(20) auto_increment,
 b int,
 primary key(a));

alter table test1 auto_increment=9223372036854775800;

insert into test1(b) values(1);

insert into test1(b) values(2);

insert into test1(b) values(3);

alter table test1 auto_increment=9223372036854775807;

alter table test1 auto_increment=9223372036854775803;

insert into test1(b) values(4);
```

### 2. What did you expect to see? (Required)
Successfully inserting `(9223372036854775803, 4)`, but failed.

Same sql done in MySQL:
```
mysql> select * from test1;
+---------------------+------+
| a                   | b    |
+---------------------+------+
| 9223372036854775800 |    1 |
| 9223372036854775801 |    2 |
| 9223372036854775802 |    3 |
| 9223372036854775803 |    4 |
| 9223372036854775804 |    5 |
| 9223372036854775805 |    6 |
+---------------------+------+
```
### 3. What did you see instead (Required)
tidb reports error:
```
ERROR 1467 (HY000): Failed to read auto-increment value from storage engine
```

### 4. What is your TiDB version? (Required)
ebdc78463a6ff52262183b67aec6c905e921be9a

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #34142 

Problem Summary:

Currently TiDB doesn't handle the overflow when allocating batches of values for `auto_increment` reaches the limitation (9223372036854775803). Only a warning would cause the following error when inserting new records.

### What is changed and how it works?

Introduce `ErrAutoIncrementOverflow` to notify the user that `auto_increment` has reached maximum limitation.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: ddl/db_integration_test.go [hunk 1]
```diff
@@ -2788,6 +2788,26 @@ func TestAutoIncrementTableOption(t *testing.T) {
...
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("drop database if exists auto_inc_overflow;")
+	tk.MustExec("create database auto_inc_overflow;")
+	tk.MustExec("use auto_inc_overflow;")
+
+	tk.MustExec("drop table if exists test1;")
+	tk.MustExec("create table test1( a bigint(20) auto_increment, b int, primary key(a));")
+	tk.MustExec("alter table test1 auto_increment=9223372036854775800;")
+	tk.MustExec("insert into test1(b) values (1),(2),(3);")
+	tk.MustGetErrCode("alter table test1 auto_increment=9223372036854775807;", errno.ErrAutoIncrementOverflow)
```
Comment (xiongjiwei) [inline ddl/db_integration_test.go:2804]: I think the alter is allowed but return a warning only.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -3277,6 +3277,9 @@ func (d *ddl) RebaseAutoID(ctx sessionctx.Context, ident ast.Ident, newBase int6
 			return err
 		}
 		if newBase != newBaseTemp {
+			if newBaseTemp < 0 {
```
Comment (xiongjiwei) [inline ddl/ddl_api.go:3280]: it seems `adjustNewBaseToNextGlobalID` will return a int64, what if the auto_increment is unsigned?

File: errno/errname.go [hunk 3]
```diff
@@ -1017,6 +1017,7 @@ var MySQLErrName = map[uint16]*mysql.ErrMessage{
 	ErrInconsistentIndexedValue:      mysql.Message("writing inconsistent data in table: %s, index: %s, col: %s, indexed-value:{%s} != record-value:{%s}", []int{3, 4}),
 	ErrAssertionFailed:               mysql.Message("assertion failed: key: %s, assertion: %s, start_ts: %v, existing start ts: %v, existing commit ts: %v", []int{0}),
 	ErrInstanceScope:                 mysql.Message("modifying %s will require SET GLOBAL in a future version of TiDB", nil),
+	ErrAutoIncrementOverflow:         mysql.Message("AUTO_INCREMENT has reached the maximum value limitation (9223372036854775807)", nil),
```
Comment (xiongjiwei) [inline errno/errname.go:1020]: the maximum value is hardcoded, should we take account into the auto_increment column type?

**Comment by tangenta** [comment]
I think that this PR does not close the issue. 

The expected behavior is that the altered ID should "take effect" when 'INSERT', rather than 'ALTER'.

File: ddl/db_integration_test.go (hunk 1 above)
Comment (buchuitoudegou) [inline ddl/db_integration_test.go:2804]: Why will `9223372036854775807` overflow? It's within the range of `bigint`.

File: errno/errcode.go [hunk 4]
```diff
@@ -1030,7 +1030,11 @@ const (
 	ErrAssertionFailed                     = 8141
 	ErrInstanceScope                       = 8142
 	ErrNonTransactionalJobFailure          = 8143
-	ErrAutoIncrementOverflow               = 8144
+	ErrSettingNoopVariable                 = 8144
+	ErrGettingNoopVariable                 = 8145
+	ErrCannotMigrateSession                = 8146
+	ErrLazyUniquenessCheckFailure          = 8147
+  ErrAutoIncrementOverflow               = 8148
```
Comment (xiongjiwei) [inline errno/errcode.go]: please fmt it

**Pull request closed without merging**